### DISCUSSION TITLE
Distorting meshes based on anisotropic resolution

### DISCUSSION THREAD

**Discussion by saskra** [JacobBumgarner/VesselVio#23] [Ideas]
**Describe the bug**
When I set the anisotropy in the visualization, neither the graph nor the labelling of the axes changes. For example, if I change 1 µm in the z-axis to 10 µm in the settings, this should also appear in the axis label and the vessels should also be ten times as long in this dimension - but they are not. (Fortunately, it seems to be only a visualization problem, at least I get different results in the analysis tab depending on the z-resolution I set.)

**To Reproduce**
Steps to reproduce the behavior: See screenshots.

**Screenshots**
![screenshot1](https://user-images.githubusercontent.com/12963485/160831086-acc2c5be-969e-4828-84d1-61476dc21ecd.png)
![screenshot2](https://user-images.githubusercontent.com/12963485/160831103-2996ff3a-ef0c-4a56-ac95-8779870054fd.png)

**Desktop (please complete the following information):**
 - OS: Windows 10

**Comment by JacobBumgarner** [comment]
I will confirm with you that this is indeed exclusively a visualization issue. The feature extraction scales all of the results based on the input resolution - that's why the scalars on the vessels will change according to the input.

At the moment, I'm only really willing to call the bounding box coordinates issue a bug. Currently I can't figure out how to do this using PyVista, but I will look into figuring out how to get this fixed! I'm gong to draw a new issue out of this.

**Comment by JacobBumgarner** [comment]
For the other aspect of this. I'm a bit unsure whether or not I want to implement this. 

When I originally decided to implement anisotropic visualization, I made the decision to only change the scalar values of the graphs and meshes. I did not want to change their coordinate positions or distort the meshes based on the resolution. This was because I wanted the vessels to still overlap with the cubic voxel inputs so that the original volume mesh could still be compared to the resulting vessel meshes.

I will say that it seems like this would be possible by calling `plotter.set_scale(x_res, y_res, z_res)`, as seen [here](https://docs.pyvista.org/api/plotting/_autosummary/pyvista.Renderer.set_scale.html?highlight=set_scale). I'm just not sure whether it makes sense to. 

I will say - I'm not completely opposed to implementing this feature. I can sit and experiment a bit with anisotropic `plotter` scaling to see what it looks like... 

Do you know how other programs similar to VesselVio deal with this problem?

**Comment by JacobBumgarner** [comment]
@saskra or @Megumi4952 - If either of you are willing, I'd love to hear some feedback on how to improve compatibility with anisotropic datasets. 

I see two possible approaches to improve compatibility:
**1.** In the first approach, the the loaded volume would be resampled into an isotropic volume based on the input resolution. This volume would then be run through the pipeline as any other normal volume. This could be achieved through affine transformations, as seen [described here](https://stackoverflow.com/questions/67888435/how-can-i-scale-an-image-so-that-it-is-isotropic).

OR

**2.** The volume is analyzed directly without resampling into an isotropic shape, but then the meshes are stretched based on the input resolution purely for visualization purposes. 

Thoughts on either of these? Or other suggestions?

I can't really tell if analyzing the raw input volume and then scaling the features based on the input resolution is kosher - thoughts? I'd be really curious to see how other programs deal with this.

**Comment by saskra** [comment]
I think you should at least display the axis markers correctly, even if you may not want/need to adjust the aspect ratio. :-) That should be easy to do, shouldn't it?

**Reply by JacobBumgarner** [reply]
Yes - the bounding box scale will be fixed in the next update 😁

**Reply by JacobBumgarner** [reply]
Just submitted a pull request for this feature over at pyvista - hopefully they'll implement it relatively soon! If not, I'll just implement it locally and then integrate their update later on.
https://github.com/pyvista/pyvista/pull/2436

**Comment by Megumi4952** [comment]
Hi Jacob - I think both make sense. Is the analysis done based on the user defined voxel size? Is the visualization independent of the analysis or user defined input? I guess in this case it would make more sense to do option 2 to avoid resampling.

**Reply by JacobBumgarner** [reply]
The structure of the visualization meshes are not dependent on the user input. However, the resulting scalar/color coding values are entirely dependent on user input. Hmmmm - ok. Definitely going to keep thinking about this one.

**Reply by saskra** [reply]
I also just double checked that the vessel length legend adjusts to the changed anisotropy settings. So it's really just a matter of how the affected axes are displayed for me.

**Comment by smith6jt-cop** [comment]
Just my two cents...

Does VesselVio rely on pixel size metadata?  If so that may explain my problems with stretching in the z dimension.  I've had plenty of hiccups related to metadata formatting compatibility between ImageJ, pyvips, tifffile, QuPath etc, but am not familiar with pyvista.

Otherwise, from what I can tell from playing around with physical size, voxel size, and input voxel size in VesselVio, the visualization relies entirely on the image dimensions for xy and the number of slices in the z dimension while, as stated, the analysis depends on the input voxel size.  If this is true, the issue I am running in to occurs when downsizing very large files for more workable file sizes.

Resizing an image down in the xy dimension results in larger scale 'xy' voxel sizes, and the easiest way to match this in the z direction is to scale the 'z' voxel size to match.  If visualization doesn't somehow pick up on this change it results in a deformed image in z.  I don't see any good use for this type of visualization.

To answer your question above, I would suggest a third option which is to have the visualization respect the voxel size supplied by the user which I assume is the actual voxel size of the image.  Why would it be helpful to have a visualization that doesn't match the measurements?  Another idea is if the image is anisotropic, prompt the user to choose isotropy as in option 1.

Also, a compromise might be to have separate inputs.  The Volume Viewer in ImageJ for example, has a 'z-Aspect' input to scale the volume visualization in the z dimension which I find helpful sometimes for exploratory purposes i.e. very dense structures can be separated.
